content/asm_6.md

```diff
@@ -1,181 +1,563 @@
+
+```assembly
+;; Calculate the dot product of the two vectors.
+_dot_product:
+        ;; Reset the value of the rax register to 0.
+        xor rax, rax
+        ;; Reset the value of the xmm1 register to 0.
+        pxor xmm1, xmm1
+        ;; Current rdx contains the number of floating-point values within the vectors.
+        ;; Multiply it by 8 to get the number of bytes occupied by these values.
+        sal rdx, 3
+;; Calculate the the dot product in the loop.
+_loop:
+        ;; Move the floating-point value from the first vector to the xmm0 register.
+        movsd xmm0, [rdi + rax]
+        ;; Multiply the floating-point from the second vector to the value from the first vector
+        ;; and store the result in the xmm0 register.
+        mulsd xmm0, [rsi + rax]
+        ;; Move to the next floating-point values in the vector buffers.
+        add rax, 8
+        ;; Add the result of multiplication of floating-point values from the vectors in the
+        ;; xmm1 register.
+        addsd xmm1, xmm0
+        ;; Check if we went through all the floating-point values in the vector buffers.
+        cmp rax, rdx
+        ;; If not yet - repeat the loop.
+        jne _loop
+        ;; Move the result to the xmm0 register.
+        movapd xmm0, xmm1
+        ;; Return from the _dot_product back to the `_calculate_dot_product`.
+        ret
+```
+
+In the beginning of the function we prepare the `rax` and `xmm1` registers by resetting them to zero. The `rax` register will contain the offset within the vector buffers and the `xmm1` will be accumulator which will accumulates result of our program.
+
+All the calculation of the dot products of our two vectors happens within the loop identified by the `_loop` label. In the first instruction after this label, we store the current value from the first vector in the register `xmm0`. After this we multiple this value by the current value from the second vector. Remember that both vectors are pointed by the `rdi` and `rsi` register and the `rax` register stores the offset within this buffers that pointer to the current value we need to process. As we did the first multiplication, we increase the value of the `rax` to eight bytes to move to the next values within the vector buffers. In the same time we update our accumulator with the intermediate result of the dot product.
 
-GNU assembler supports following operators for far functions call and jumps:
+At this point, the `xmm1` register will have the result of multiplication of the fist components of our vectors. At the next step we check that - did we reach the end of vectors and if not we repeat the loop for the second components, third, and so on.
+
```

Comment:
I would remove this line break and join these paragraphs into one

Reply:
Done